homeassistant/components/zwave_js/climate.py

```diff
@@ -0,0 +1,303 @@
+            add_to_watched_value_ids=True,
+        )
+
+        self.on_value_update()
+
+    @callback
+    def on_value_update(self) -> None:
+        """Call when a watched value is added or updated."""
+        self._current_mode_setpoint_values = self._get_current_mode_setpoint_values()
+
+    def _get_current_mode_setpoint_values(self) -> List[ZwaveValue]:
+        """Return a list of current setpoint Z-Wave value(s)."""
+        current_mode = int(self.info.primary_value.value)
+        setpoint_enums: List[ThermostatSetpointType] = THERMOSTAT_MODE_SETPOINT_MAP.get(
+            current_mode, []
+        )
+        # we do not want None values in our list so check if the value exists
+        values = []
+        for enum in setpoint_enums:
+            value: Optional[ZwaveValue] = self.get_zwave_value(
+                THERMOSTAT_SETPOINT_PROPERTY,
+                CommandClass.THERMOSTAT_SETPOINT,
+                value_property_key_name=enum.value,
+            )
+            if value:
+                values.append(value)
+        return values
+
+    def _set_modes_and_presets(self) -> None:
+        """Convert Z-Wave Thermostat modes into Home Assistant modes and presets."""
+        all_modes: Dict[str, Optional[int]] = {}
+        all_presets: Dict[str, Optional[int]] = {PRESET_NONE: None}
+
+        # Z-Wave uses one list for both modes and presets.
+        # Iterate over all Z-Wave ThermostatModes and extract the hvac modes and presets.
+        for mode_id, mode_name in self.info.primary_value.metadata.states.items():
+            mode_id = int(mode_id)
+            if mode_id in THERMOSTAT_MODES:
+                # treat value as hvac mode
+                hass_mode = ZW_HVAC_MODE_MAP.get(mode_id)
+                if hass_mode:
+                    all_modes[hass_mode] = mode_id
+            else:
+                # treat value as hvac preset
+                all_presets[mode_name] = mode_id
+        self._hvac_modes = all_modes
+        self._hvac_presets = all_presets
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement used by the platform."""
+        temp: Optional[ZwaveValue] = None
+        if self._current_mode_setpoint_values:
+            temp = self._current_mode_setpoint_values[0]
+        if temp is not None and temp.metadata.unit == "°F":
+            return TEMP_FAHRENHEIT
+        return TEMP_CELSIUS
+
+    @property
+    def hvac_mode(self) -> str:
+        """Return hvac operation ie. heat, cool mode."""
+        return ZW_HVAC_MODE_MAP.get(
+            int(self.info.primary_value.value), HVAC_MODE_HEAT_COOL
+        )
+
+    @property
+    def hvac_modes(self) -> List[str]:
+        """Return the list of available hvac operation modes."""
+        return list(self._hvac_modes)
+
+    @property
+    def hvac_action(self) -> Optional[str]:
+        """Return the current running hvac operation if supported."""
+        operating_state = self.get_zwave_value(
```

Comment:
We should be able to store the operating state value as an attribute when setting up the entity.

Reply:
here's where maybe I am getting confused by these comments: as far as I can tell, the Value object gets replaced on every update, so storing the Value wouldn't be useful. Am I misunderstanding that? Does the Value object get updated on updates instead of replaced?

Reply:
Yes, the Value instance remains. It's just its data that is updated.

https://github.com/home-assistant-libs/zwave-js-server-python/blob/3e996393d940fa9806fcc7eeb85e60c29328b000/zwave_js_server/model/node.py#L316-L332

https://github.com/home-assistant-libs/zwave-js-server-python/blob/3e996393d940fa9806fcc7eeb85e60c29328b000/zwave_js_server/model/value.py#L169

Reply:
🤦🏾 I missed the if statement at the beginning and assumed event updates always created new values. I understand now, thanks for walking me through this